File: Avail/src/com/avail/descriptor/AtomWithPropertiesDescriptor.java
/**
 * AtomWithPropertiesDescriptor.java
 * Copyright © 1993-2013, Mark van Gulik and Todd L Smith.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of the copyright holder nor the names of the contributors
 *   may be used to endorse or promote products derived from this software
 *   without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

package com.avail.descriptor;

import static com.avail.descriptor.AtomWithPropertiesDescriptor.IntegerSlots.*;
import static com.avail.descriptor.AtomWithPropertiesDescriptor.ObjectSlots.*;
import com.avail.annotations.*;
import com.avail.serialization.Serializer;

/**
 * An {@code atom} is an object that has identity by fiat, i.e., it is
 * distinguished from all other objects by the fact of its creation event and
 * the history of what happens to its references.  Not all objects in Avail have
 * that property (hence the acronym Advanced Value And Identity Language),
 * unlike most object-oriented programming languages.
 *
 * <p>
 * At any time an atom can have properties associated with it.  A property is
 * an association between another atom, known as the property key, and the value
 * of that property, any Avail object.  Atoms without properties have a
 * {@linkplain AtomDescriptor representation} that does not include a slot for
 * the properties information, but adding a property causes it to transform (via
 * {@link AvailObject#becomeIndirectionTo(A_BasicObject)} into a {@linkplain
 * AtomWithPropertiesDescriptor representation} that has a slot which contains
 * a map from property keys to property values.
 * </p>
 *
 * @author Mark van Gulik &lt;mark@availlang.org&gt;
 * @author Todd L Smith &lt;todd@availlang.org&gt;
 * @see AtomDescriptor
 * @see AtomWithPropertiesSharedDescriptor
 */
public class AtomWithPropertiesDescriptor
extends AtomDescriptor
{
	/**
	 * The layout of integer slots for my instances.
	 */
	public enum IntegerSlots
	implements IntegerSlotsEnum
	{
		/**
		 * The hash value of this {@linkplain AtomDescriptor atom}.  It is a
		 * random number (not 0), computed on demand.
		 */
		@HideFieldInDebugger
		HASH_OR_ZERO;

		static
		{
			assert AtomDescriptor.IntegerSlots.HASH_OR_ZERO.ordinal()
				== HASH_OR_ZERO.ordinal();
		}
	}

	/**
	 * The layout of object slots for my instances.
	 */
	public enum ObjectSlots
	implements ObjectSlotsEnum
	{
		/**
		 * A string (non-uniquely) roughly identifying this atom.  It need not
		 * be unique among atoms.  Must have the same ordinal as {@link
		 * AtomDescriptor.ObjectSlots#NAME}.
		 */
		NAME,

		/**
		 * The {@linkplain ModuleDescriptor module} that was active when this
		 * atom was issued.  This information is crucial to {@linkplain
		 * Serializer serialization}.  Must have the same ordinal as
		 * {@link AtomDescriptor.ObjectSlots#NAME}.
		 */
		ISSUING_MODULE,

		/**
		 * A map from this atom's property keys (atoms) to property values.
		 */
		PROPERTY_MAP;

		static
		{
			assert AtomDescriptor.ObjectSlots.NAME.ordinal()
				== NAME.ordinal();
			assert AtomDescriptor.ObjectSlots.ISSUING_MODULE.ordinal()
				== ISSUING_MODULE.ordinal();
		}
	}

	@Override
	boolean allowsImmutableToMutableReferenceInField (final AbstractSlotsEnum e)
	{
		return super.allowsImmutableToMutableReferenceInField(e)
			|| e == HASH_OR_ZERO
			|| e == PROPERTY_MAP;
	}

	@Override
	AvailObject o_MakeShared (final AvailObject object)
	{
		assert !isShared();
		// The layout of the destination descriptor is the same, so nothing
		// special needs to happen, i.e., object doesn't need to become an
		// indirection.
		object.descriptor = AtomWithPropertiesSharedDescriptor.shared;
		object.slot(PROPERTY_MAP).makeShared();
		return object;
	}

	/**
	 * {@inheritDoc}
	 *
	 * <p>
	 * Add or replace a property of this {@linkplain AtomDescriptor atom with
	 * properties}.
	 * </p>
	 */
	@Override @AvailMethod
	void o_SetAtomProperty (
		final AvailObject object,
		final A_Atom key,
		final A_BasicObject value)
	{
		assert key.isAtom();
		A_Map map = object.slot(PROPERTY_MAP);
		if (value.equalsNil())
		{
			map = map.mapWithoutKeyCanDestroy(key, true);
		}
		else
		{
			map = map.mapAtPuttingCanDestroy(key, value, true);
		}
		if (isShared())
		{
			map = map.traversed().makeShared();
		}
		object.setSlot(PROPERTY_MAP, map);
	}

	/**
	 * {@inheritDoc}
	 *
	 * <p>
	 * Extract the property value of this atom at the specified key.  Return
	 * {@linkplain NilDescriptor#nil() nil} if no such
	 * property exists.
	 * </p>
	 */
	@Override @AvailMethod
	AvailObject o_GetAtomProperty (
		final AvailObject object,
		final A_Atom key)
	{
		assert key.isAtom();
		final A_Map map = object.slot(PROPERTY_MAP);
		if (map.hasKey(key))
		{
			return map.mapAt(key);
		}
		return NilDescriptor.nil();
	}

	/**
	 * Create a new atom with the given name.  The name is not globally unique,
	 * but serves to help to visually distinguish atoms.  In this class, the
	 * created object already has an empty property map.
	 *
	 * @param name
	 *            A string used to help identify the new atom.
	 * @param issuingModule
	 *            Which {@linkplain ModuleDescriptor module} was active when the
	 *            atom was created.
	 * @return
	 *            The new atom, not equal to any object in use before this
	 *            method was invoked.
	 */
	public static AvailObject create (
		final AvailObject name,
		final AvailObject issuingModule)
	{
		final AvailObject instance = mutable.create();
		instance.setSlot(NAME, name);
		instance.setSlot(ISSUING_MODULE, issuingModule);
		instance.setSlot(PROPERTY_MAP, MapDescriptor.empty());
		instance.setSlot(HASH_OR_ZERO, 0);
		instance.makeImmutable();
		return instance;
	}

	/**
	 * Create a new atom with the given name, module, and hash value.  The name
	 * is not globally unique, but serves to help to visually distinguish atoms.
	 * The hash value is provided to allow an existing {@linkplain
	 * AtomDescriptor simple atom} to be converted to an {@linkplain
	 * AtomWithPropertiesDescriptor atom with properties}.  The client can
	 * convert the original simple atom into an {@linkplain
	 * IndirectionDescriptor indirection} to the new atom with properties.
	 *
	 * @param name
	 *            A string used to help identify the new atom.
	 * @param issuingModule
	 *            The module that issued this atom.
	 * @param originalHash
	 *            The hash value that must be set for this atom, or zero if it
	 *            doesn't matter.
	 * @return
	 *            The new atom, not equal to any object in use before this
	 *            method was invoked.
	 */
	public static AvailObject createWithNameAndModuleAndHash (
		final AvailObject name,
		final AvailObject issuingModule,
		final int originalHash)
	{
		final AvailObject instance = mutable.create();
		instance.setSlot(NAME, name);
		instance.setSlot(ISSUING_MODULE, issuingModule);
		instance.setSlot(PROPERTY_MAP, MapDescriptor.empty());
		instance.setSlot(HASH_OR_ZERO, originalHash);
		instance.makeImmutable();
		return instance;
	}

	/**
	 * Construct a new {@link AtomWithPropertiesDescriptor}.
	 *
	 * @param mutability
	 *        The {@linkplain Mutability mutability} of the new descriptor.
	 */
	protected AtomWithPropertiesDescriptor (final Mutability mutability)
	{
		super(mutability);
	}

	/** The mutable {@link AtomWithPropertiesDescriptor}. */
	private static final AtomWithPropertiesDescriptor mutable =
		new AtomWithPropertiesDescriptor(Mutability.MUTABLE);

	@Override
	AtomWithPropertiesDescriptor mutable ()
	{
		return mutable;
	}

	/** The immutable {@link AtomWithPropertiesDescriptor}. */
	private static final AtomWithPropertiesDescriptor immutable =
		new AtomWithPropertiesDescriptor(Mutability.IMMUTABLE);

	@Override
	AtomWithPropertiesDescriptor immutable ()
	{
		return immutable;
	}
}

File: avail-core/src/main/kotlin/avail/descriptor/atoms/AtomWithPropertiesDescriptor.kt
/*
 * AtomWithPropertiesDescriptor.kt
 * Copyright © 1993-2021, The Avail Foundation, LLC.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * * Neither the name of the copyright holder nor the names of the contributors
 *   may be used to endorse or promote products derived from this software
 *   without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
package avail.descriptor.atoms

import avail.annotations.HideFieldInDebugger
import avail.descriptor.atoms.A_Atom.Companion.atomName
import avail.descriptor.atoms.A_Atom.Companion.getAtomProperty
import avail.descriptor.atoms.AtomDescriptor.SpecialAtom.EXPLICIT_SUBCLASSING_KEY
import avail.descriptor.atoms.AtomDescriptor.SpecialAtom.HERITABLE_KEY
import avail.descriptor.atoms.AtomWithPropertiesDescriptor.IntegerSlots.Companion.HASH_OR_ZERO
import avail.descriptor.atoms.AtomWithPropertiesDescriptor.IntegerSlots.HASH_AND_MORE
import avail.descriptor.atoms.AtomWithPropertiesDescriptor.ObjectSlots.ISSUING_MODULE
import avail.descriptor.atoms.AtomWithPropertiesDescriptor.ObjectSlots.NAME
import avail.descriptor.atoms.AtomWithPropertiesDescriptor.ObjectSlots.PROPERTY_MAP_POJO
import avail.descriptor.module.A_Module
import avail.descriptor.module.ModuleDescriptor
import avail.descriptor.pojos.RawPojoDescriptor.Companion.identityPojo
import avail.descriptor.representation.A_BasicObject
import avail.descriptor.representation.AbstractSlotsEnum
import avail.descriptor.representation.AvailObject
import avail.descriptor.representation.AvailObjectFieldHelper
import avail.descriptor.representation.BitField
import avail.descriptor.representation.IndirectionDescriptor
import avail.descriptor.representation.IntegerSlotsEnum
import avail.descriptor.representation.Mutability
import avail.descriptor.representation.NilDescriptor.Companion.nil
import avail.descriptor.representation.ObjectSlotsEnum
import avail.descriptor.tuples.A_String
import avail.descriptor.types.TypeTag
import avail.serialization.Serializer
import avail.serialization.SerializerOperation
import java.util.WeakHashMap

/**
 * An `atom` is an object that has identity by fiat, i.e., it is distinguished
 * from all other objects by the fact of its creation event and the history of
 * what happens to its references.  Not all objects in Avail have that property
 * (hence the acronym Advanced Value And Identity Language), unlike most
 * object-oriented programming languages.
 *
 * At any time an atom can have properties associated with it.  A property is an
 * association between another atom, known as the property key, and the value of
 * that property, any Avail object.  Atoms without properties have a
 * [representation][AtomDescriptor] that does not include a slot for the
 * properties information, but adding a property causes it to transform (via
 * [AvailObject.becomeIndirectionTo] into an [AtomWithPropertiesDescriptor]
 * representation that has a slot which contains a map from property keys to
 * property values.
 *
 * @constructor
 *
 * @param mutability
 *   The [mutability][Mutability] of the new descriptor.
 * @param typeTag
 *   The [TypeTag] to use in this descriptor.
 * @param objectSlotsEnumClass
 *   The Java [Class] which is a subclass of [ObjectSlotsEnum] and defines this
 *   object's object slots layout, or null if there are no object slots.
 * @param integerSlotsEnumClass
 *   The Java [Class] which is a subclass of [IntegerSlotsEnum] and defines this
 *   object's integer slots layout, or null if there are no integer slots.
 *
 * @author Mark van Gulik &lt;mark@availlang.org&gt;
 * @author Todd L Smith &lt;todd@availlang.org&gt;
 *
 * @see AtomDescriptor
 * @see AtomWithPropertiesSharedDescriptor
 */
open class AtomWithPropertiesDescriptor protected constructor(
	mutability: Mutability,
	typeTag: TypeTag,
	objectSlotsEnumClass: Class<out ObjectSlotsEnum>,
	integerSlotsEnumClass: Class<out IntegerSlotsEnum>
) : AtomDescriptor(
	mutability, typeTag, objectSlotsEnumClass, integerSlotsEnumClass)
{
	/**
	 * The layout of integer slots for my instances.
	 */
	enum class IntegerSlots : IntegerSlotsEnum {
		/**
		 * The low 32 bits are used for the [HASH_OR_ZERO], but the upper 32 can
		 * be used by other [BitField]s in subclasses.
		 */
		@HideFieldInDebugger
		HASH_AND_MORE;

		companion object {
			/**
			 * A slot to hold the hash value, or zero if it has not been
			 * computed. The hash of an atom is a random number, computed once.
			 */
			val HASH_OR_ZERO = BitField(HASH_AND_MORE, 0, 32) { null }

			init {
				assert(AtomDescriptor.IntegerSlots.HASH_AND_MORE.ordinal
					== HASH_AND_MORE.ordinal)
				assert(AtomDescriptor.IntegerSlots.HASH_OR_ZERO.isSamePlaceAs(
					HASH_OR_ZERO))
			}
		}
	}

	/**
	 * The layout of object slots for my instances.
	 */
	enum class ObjectSlots : ObjectSlotsEnum {
		/**
		 * A string (non-uniquely) roughly identifying this atom.  It need not
		 * be unique among atoms.  Must have the same ordinal as
		 * [AtomDescriptor.ObjectSlots.NAME].
		 */
		NAME,

		/**
		 * The [module][ModuleDescriptor] that was active when this atom was
		 * issued.  This information is crucial to [serialization][Serializer].
		 * Must have the same ordinal as [AtomDescriptor.ObjectSlots.NAME].
		 */
		ISSUING_MODULE,

		/**
		 * A pojo holding a weak map from this atom's property keys (atoms) to
		 * property values.  It's never [nil] for this descriptor class, but it
		 * may be [nil] in the [Mutability.SHARED] subclass.
		 */
		@HideFieldInDebugger
		PROPERTY_MAP_POJO;

		companion object {
			init {
				assert(AtomDescriptor.ObjectSlots.NAME.ordinal
					== NAME.ordinal)
				assert(AtomDescriptor.ObjectSlots.ISSUING_MODULE.ordinal
					== ISSUING_MODULE.ordinal)
			}
		}
	}

	override fun allowsImmutableToMutableReferenceInField(
		e: AbstractSlotsEnum
	) = super.allowsImmutableToMutableReferenceInField(e)
		|| e === HASH_AND_MORE

	override fun o_DescribeForDebugger(
		self: AvailObject
	): Array<AvailObjectFieldHelper>
	{
		val fieldsArray = super.o_DescribeForDebugger(self)
		val propertiesPojo = self.slot(PROPERTY_MAP_POJO)
		if (propertiesPojo.isNil)
		{
			return fieldsArray
		}
		val fields = fieldsArray.toMutableList()
		val properties =
			propertiesPojo.javaObjectNotNull<Map<A_Atom, A_BasicObject>>()
		properties.entries
			.sortedBy { it.key.nameForDebugger() }
			.forEach { (key, value) ->
				fields.add(
					AvailObjectFieldHelper(
						self,
						DebuggerObjectSlots.DUMMY_DEBUGGER_SLOT,
						-1,
						value,
						slotName = "Key " + key.atomName))
			}
		return fields.toTypedArray()
	}

	/**
	 * Extract the property value of this atom at the specified key.  Return
	 * [nil] if no such property exists.
	 */
	override fun o_GetAtomProperty(
		self: AvailObject,
		key: A_Atom
	): AvailObject {
		assert(key.isAtom)
		val propertyMapPojo: A_BasicObject = self.slot(PROPERTY_MAP_POJO)
		val propertyMap: Map<A_Atom, AvailObject> =
			propertyMapPojo.javaObjectNotNull()
		return propertyMap[key] ?: nil
	}

	override fun o_MakeShared(self: AvailObject): AvailObject {
		assert(!isShared)
		val propertyMapPojo = self.slot(PROPERTY_MAP_POJO)
		assert(propertyMapPojo.notNil)
		val substituteAtom: AvailObject =
			AtomWithPropertiesSharedDescriptor.shared.createInitialized(
				self.slot(NAME),
				self.slot(ISSUING_MODULE),
				propertyMapPojo,
				self.slot(HASH_OR_ZERO))
		self.becomeIndirectionTo(substituteAtom)
		// In case a property key is the atom itself, only make the property
		// keys and atoms shared after making self shared.  Otherwise there
		// could be unbounded recursion.
		val propertyMap: Map<A_Atom, AvailObject> =
			propertyMapPojo.javaObjectNotNull()
		propertyMap.forEach { (key, value) ->
			key.makeShared()
			value.makeShared()
		}
		return substituteAtom
	}

	override fun o_SerializerOperation (self: AvailObject) = when {
		self.getAtomProperty(HERITABLE_KEY.atom).notNil ->
			SerializerOperation.HERITABLE_ATOM
		self.getAtomProperty(EXPLICIT_SUBCLASSING_KEY.atom).notNil ->
			SerializerOperation.EXPLICIT_SUBCLASS_ATOM
		else -> SerializerOperation.ATOM
	}

	/**
	 * Add or replace a property of this [A_Atom].  If the provided value is
	 * [nil], remove the property.
	 */
	override fun o_SetAtomProperty(
		self: AvailObject,
		key: A_Atom,
		value: A_BasicObject
	) {
		assert(key.isAtom)
		val propertyMapPojo = self.slot(PROPERTY_MAP_POJO)
		val map: MutableMap<A_Atom, A_BasicObject> =
			propertyMapPojo.javaObjectNotNull()
		when {
			value.isNil -> map.remove(key)
			else -> map[key.makeShared()] = value.makeShared()
		}
	}

	override fun mutable() = mutable

	override fun immutable() = immutable

	companion object {
		/**
		 * Create a new atom with the given name, module, and hash value.  The
		 * name is not globally unique, but serves to help to visually
		 * distinguish atoms. The hash value is provided to allow an existing
		 * [simple&#32;atom][AtomDescriptor] to be converted to an
		 * [atom&#32;with&#32;properties][AtomWithPropertiesDescriptor].  The
		 * client can convert the original simple atom into an
		 * [indirection][IndirectionDescriptor] to the new atom with properties.
		 *
		 * @param name
		 *   An [A_String] used to help identify the new atom.
		 * @param issuingModule
		 *   The [A_Module] that issued this atom.
		 * @param originalHashOrZero
		 *   The hash value that must be set for this atom, or zero if it has
		 *   not yet been computed.
		 * @return
		 *   The new atom, not equal to any object in use before this method was
		 *   invoked.
		 */
		fun createWithProperties(
			name: A_String,
			issuingModule: A_Module,
			originalHashOrZero: Int
		): AvailObject = mutable.create {
			setSlot(NAME, name)
			setSlot(ISSUING_MODULE, issuingModule)
			setSlot(
				PROPERTY_MAP_POJO,
				identityPojo(WeakHashMap<A_Atom, A_BasicObject>()))
			setSlot(HASH_OR_ZERO, originalHashOrZero)
		}

		/** The mutable [AtomWithPropertiesDescriptor]. */
		private val mutable = AtomWithPropertiesDescriptor(
			Mutability.MUTABLE,
			TypeTag.ATOM_TAG,
			ObjectSlots::class.java,
			IntegerSlots::class.java)

		/** The immutable [AtomWithPropertiesDescriptor]. */
		private val immutable = AtomWithPropertiesDescriptor(
			Mutability.IMMUTABLE,
			TypeTag.ATOM_TAG,
			ObjectSlots::class.java,
			IntegerSlots::class.java)
	}
}
